src/Compilers/Core/Portable/Operations/ControlFlowGraphBuilder.cs

```diff
@@ -0,0 +1,555 @@
+﻿// Copyright (c) Microsoft.  All Rights Reserved.  Licensed under the Apache License, Version 2.0.  See License.txt in the project root for license information.
+
+using System;
+using System.Collections.Concurrent;
+using System.Collections.Immutable;
+using System.Diagnostics;
+using System.Linq;
+using Microsoft.CodeAnalysis.PooledObjects;
+using Roslyn.Utilities;
+
+namespace Microsoft.CodeAnalysis.Operations
+{
+    internal sealed class ControlFlowGraphBuilder : OperationCloner
+    {
+        private readonly BasicBlock _entry = new BasicBlock(BasicBlockKind.Entry);
+        private readonly BasicBlock _exit = new BasicBlock(BasicBlockKind.Exit);
+        private ArrayBuilder<BasicBlock> _blocks;
+        private BasicBlock _currentBasicBlock;
+        private IOperation _currentStatement;
+        private ArrayBuilder<IOperation> _evalStack;
+        private int _availableCaptureId;
+
+        private ControlFlowGraphBuilder()
+        { }
+
+        public static ImmutableArray<BasicBlock> Create(IBlockOperation body)
```

Comment:
Should we consider caching the result on the semantic model?

Reply:
Why? This API is not tied to any SemanticModel instance. Consumer can cache or not cache the graph as appropriate. 

---
In reply to: [162195775](https://github.com/dotnet/roslyn/pull/24263#discussion_r162195775) [](ancestors = 162195775)